local development error from **make run** main: error: Cannot start controller manager: open /tls/server/tls.crt: no such file or directory

### What happened?
We are missing the tls path that is expected when running controller locally. 
```
🦄 ❯❯❯ make run 
10:38:52 [ .. ] verify go modules dependencies have expected content
all modules verified
10:39:02 [ OK ] go modules dependencies verified
10:39:02 [ .. ] go generate linux_arm64
10:39:07 [ OK ] go generate linux_arm64
10:39:07 [ .. ] go mod tidy
10:39:07 [ OK ] go mod tidy
10:39:07 [ .. ] Generating CRDs with kustomize
10:39:08 [ OK ] Generated CRDs with kustomize
10:39:08 [ .. ] Cleaning up kustomize generated CRDs
10:39:08 [ OK ] Cleaned up kustomize generated CRDs
10:39:08 [ .. ] Running Crossplane locally out-of-cluster . . .
customresourcedefinition.apiextensions.k8s.io/objects.kubernetes.crossplane.io unchanged
customresourcedefinition.apiextensions.k8s.io/observedobjectcollections.kubernetes.crossplane.io unchanged
customresourcedefinition.apiextensions.k8s.io/providerconfigs.kubernetes.crossplane.io unchanged
customresourcedefinition.apiextensions.k8s.io/providerconfigusages.kubernetes.crossplane.io unchanged
go run cmd/provider/main.go -d
2024-05-21T10:39:11.395+0200	DEBUG	provider-kubernetes	Starting	{"sync-interval": "1h0m0s", "poll-interval": "10m0s", "poll-jitter": "1m0s", "max-reconcile-rate": 100}
2024-05-21T10:39:11.398+0200	INFO	provider-kubernetes	Beta feature enabled	{"flag": "EnableBetaManagementPolicies"}
main: error: Cannot start controller manager: open /tls/server/tls.crt: no such file or directory
exit status 1
make: *** [run] Error 1
```

### How can we reproduce it?
run make target `make run` from main

--- a/Makefile
+++ b/Makefile
@@ -119,10 +119,17 @@ build.init: $(CROSSPLANE_CLI)
 # This is for running out-of-cluster locally, and is for convenience. Running
 # this make target will print out the command which was used. For more control,
 # try running the binary directly with different arguments.
-run: $(KUBECTL) generate
+# The objects CRD ships a partial conversion-webhook stanza that only the
+# Crossplane package manager can complete at install time, so it is stripped
+# before applying, and the provider runs with an empty --certs-dir, which
+# disables the conversion webhook so that no TLS certificates are needed.
+# See #343 and #249.
+run: $(KUBECTL) $(YQ) generate
 	@$(INFO) Running Crossplane locally out-of-cluster . . .
-	@$(KUBECTL) apply -f package/crds/ -R
-	go run cmd/provider/main.go -d
+	@for crd in package/crds/*.yaml; do \
+		$(YQ) 'del(.spec.conversion)' "$$crd" | $(KUBECTL) apply -f - || exit 1; \
+	done
+	go run cmd/provider/main.go -d --certs-dir=""
 
 manifests:
 	@$(INFO) Deprecated. Run make generate instead.

--- a/cmd/provider/main.go
+++ b/cmd/provider/main.go
@@ -74,6 +74,7 @@ import (
 const (
 	webhookTLSCertDirEnvVar = "WEBHOOK_TLS_CERT_DIR"
 	tlsServerCertDirEnvVar  = "TLS_SERVER_CERTS_DIR"
+	certsDirEnvVar          = "CERTS_DIR"
 	tlsServerCertDir        = "/tls/server"
 )
 
@@ -109,6 +110,14 @@ func main() {
 		// Additional legacy field managers to upgrade to Server-side apply field manager
 		legacyCSAFieldManagers = app.Flag("legacy-csa-field-managers", "Additional legacy client-side apply Kubernetes field manager names for upgrading to SSA field manager").Default().Strings()
 		enableSecretCache      = app.Flag("enable-secret-cache", "Enable caching of Secret objects. When true, Secrets are served from the informer cache instead of direct API calls. This reduces API server load but increases memory usage.").Default("true").Envar("ENABLE_SECRET_CACHE").Bool()
+
+		certsDirSet = false
+		// we record whether the command-line option "--certs-dir" was supplied
+		// in the registered PreAction for the flag.
+		certsDir = app.Flag("certs-dir", "The directory that contains the server key and certificate. Set to an empty string to disable the conversion webhook, e.g. when running out-of-cluster without TLS certificates.").Default(tlsServerCertDir).Envar(certsDirEnvVar).PreAction(func(_ *kingpin.ParseContext) error {
+			certsDirSet = true
+			return nil
+		}).String()
 	)
 	kingpin.MustParse(app.Parse(os.Args[1:]))
 
@@ -136,15 +145,25 @@ func main() {
 	cfg, err := ctrl.GetConfig()
 	kingpin.FatalIfError(err, "Cannot get API server rest config")
 
-	// Get the TLS certs directory from the environment variable if set
+	// Get the TLS certs directory from the environment variables set by
+	// Crossplane if they're available.
 	// In older XP versions we used WEBHOOK_TLS_CERT_DIR, in newer versions
-	// we use TLS_SERVER_CERTS_DIR. If neither are set, use the default.
-	var certDir string
-	certDir = os.Getenv(webhookTLSCertDirEnvVar)
-	if certDir == "" {
-		certDir = os.Getenv(tlsServerCertDirEnvVar)
-		if certDir == "" {
-			certDir = tlsServerCertDir
+	// we use TLS_SERVER_CERTS_DIR. If an explicit certs dir is not supplied
+	// via the command-line options, then these environment variables are used
+	// instead.
+	if !certsDirSet {
+		// backwards-compatibility concerns
+		xpCertsDir := os.Getenv(certsDirEnvVar)
+		if xpCertsDir == "" {
+			xpCertsDir = os.Getenv(tlsServerCertDirEnvVar)
+		}
+		if xpCertsDir == "" {
+			xpCertsDir = os.Getenv(webhookTLSCertDirEnvVar)
+		}
+		// we probably don't need this condition but just to be on the
+		// safe side, if we are missing any kingpin machinery details...
+		if xpCertsDir != "" {
+			*certsDir = xpCertsDir
 		}
 	}
 
@@ -193,13 +212,12 @@ func main() {
 		LeaseDuration:              func() *time.Duration { d := 60 * time.Second; return &d }(),
 		RenewDeadline:              func() *time.Duration { d := 50 * time.Second; return &d }(),
 		WebhookServer: webhook.NewServer(webhook.Options{
-			CertDir: certDir,
+			CertDir: *certsDir,
 			Port:    *webhookPort,
 		}),
 		HealthProbeBindAddress: *healthProbeBindAddress,
 	})
 	kingpin.FatalIfError(err, "Cannot create controller manager")
-	kingpin.FatalIfError(mgr.AddReadyzCheck("webhook", mgr.GetWebhookServer().StartedChecker()), "Cannot add webhook server readyz checker to controller manager")
 
 	// The manager client reads through informers that only start with the
 	// manager, so the startup listing goes straight to the API server.
@@ -209,6 +227,10 @@ func main() {
 	cacheSize, err := clientCacheSize(sizingCtx, log, sizingClient)
 	cancelSizing()
 	kingpin.FatalIfError(err, "Cannot calculate the client cache size")
+
+	if len(*certsDir) > 0 {
+		kingpin.FatalIfError(mgr.AddReadyzCheck("webhook", mgr.GetWebhookServer().StartedChecker()), "Cannot add webhook server readyz checker to controller manager")
+	}
 
 	mm := managed.NewMRMetricRecorder()
 	sm := statemetrics.NewMRStateMetrics()
@@ -275,11 +297,13 @@ func main() {
 		o.ChangeLogOptions = &clo
 	}
 
-	// NOTE(lsviben): We are registering the conversion webhook with v1alpha1
-	// Object. As far as I can see and based on some tests, it doesn't matter
-	// which version we use here. Leaving it as v1alpha1 as it will be easy to
-	// notice and remove when we drop support for v1alpha1.
-	kingpin.FatalIfError(ctrl.NewWebhookManagedBy(mgr, &objectv1alpha1cluster.Object{}).Complete(), "Cannot create Object webhook") //nolint:staticcheck // registering conversion webhook for deprecated api
+	if len(*certsDir) > 0 {
+		// NOTE(lsviben): We are registering the conversion webhook with v1alpha1
+		// Object. As far as I can see and based on some tests, it doesn't matter
+		// which version we use here. Leaving it as v1alpha1 as it will be easy to
+		// notice and remove when we drop support for v1alpha1.
+		kingpin.FatalIfError(ctrl.NewWebhookManagedBy(mgr, &objectv1alpha1cluster.Object{}).Complete(), "Cannot create Object webhook") //nolint:staticcheck // registering conversion webhook for deprecated api
+	}
 	precheckCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 	defer cancel()
 	canSafeStart, err := canWatchCRD(precheckCtx, mgr)

